packages/cli/src/controller/generate-controller.ts

```diff
@@ -0,0 +1,317 @@
+  });
+
+  // for (const key in availableFragments) {
+  //   if (inputs.includes(availableFragments[key].name.toLowerCase())) {
+  //     selectedFragments[key] = availableFragments[key];
+  //   }
+  //   //   console.log(chalk.red(`"${matchingInputs}" are duplicated on ABI: ${abiName}`));
+  // }
+  return selectedFragments;
+}
+
+export function filterExistingMethods(
+  eventFragments: Record<string, EventFragment>,
+  functionFragments: Record<string, FunctionFragment>,
+  dataSources: EthereumDs[]
+): [Record<string, EventFragment>, Record<string, FunctionFragment>] {
+  // find all existing functions and events
+  const existingEvents: string[] = [];
+  const existingFunctions: string[] = [];
+
+  const cleanEventsFragments: Record<string, EventFragment> = {};
+  const cleanFunctionsFragments: Record<string, FunctionFragment> = {};
+
+  dataSources.map((ds) => {
+    ds.mapping.handlers.map((handler) => {
+      if (Object.keys(handler.filter).includes('topics')) {
+        // topic[0] is the method
+        existingEvents.push((handler.filter as EthereumLogFilter).topics[0]);
+      }
+      if (Object.keys(handler.filter).includes('function')) {
+        existingFunctions.push((handler.filter as EthereumTransactionFilter).function);
+      }
+    });
+  });
+
+  for (const key in eventFragments) {
+    const fragmentFormats = Object.values(getFragmentFormats<EventFragment>(eventFragments[key])).concat(key);
+    const diff = difference(fragmentFormats, existingEvents);
+    // console.log(fragmentFormats, existingEvents)
+    if (diff.length === fragmentFormats.length) {
+      diff.map((fragKey) => {
+        if (eventFragments[fragKey]) {
+          cleanEventsFragments[fragKey] = eventFragments[fragKey];
+        }
+      });
+    }
+  }
+  for (const key in functionFragments) {
+    const fragmentFormats = Object.values(getFragmentFormats<FunctionFragment>(functionFragments[key])).concat(key);
+    const diff = difference(fragmentFormats, existingFunctions);
+    if (diff.length === fragmentFormats.length) {
+      diff.map((fragKey) => {
+        if (functionFragments[fragKey]) {
+          cleanFunctionsFragments[fragKey] = functionFragments[fragKey];
+        }
+      });
+    }
+  }
+  return [cleanEventsFragments, cleanFunctionsFragments];
+}
+
+export async function getManifestData(projectPath: string, manifestPath: string): Promise<any> {
+  const existingManifest = (await fs.promises.readFile(path.join(projectPath, manifestPath), 'utf8')) as any;
+  return parseDocument(existingManifest);
+}
+
+export async function generateManifest(
+  projectPath: string,
+  manifestPath: string,
+  userInput: UserInput,
+  existingManifestData: any
+): Promise<void> {
+  try {
+    const clonedExistingManifestData = existingManifestData.clone();
+    const existingDatasource = existingManifestData.get('dataSources') as any;
+
+    const newDataSourcesData = existingDatasource.toJSON().concat(...[constructDatasources(userInput)]);
```

Comment:
I think we're going to lose any comments in the datasources portion by using `toJSON`